File: ADT/trees/501.cpp
// Google

// Tc = O(n)
// Sc = O(n)

class Solution {
public:
    unordered_map<int, int> mp;
    void dfs(TreeNode* root) {
        if (!root) {
            return;
        }
        dfs(root->left);
        mp[root->val]++;
        dfs(root->right);
    }
    vector<int> findMode(TreeNode* root) {
        dfs(root);
        vector<int> result;
        int maxFreq = 0;
        for (auto& it : mp) {
            if (it.second > maxFreq) {
                maxFreq = it.second;
                result = {};
                result.push_back(it.first);
            } else if (it.second ==  maxFreq) {
                result.push_back(it.first);
            }
        }
        return result;
    }
};



// Follow-up
// Tc = O(n)
// Sc = O(1)


class Solution {
public:
    int currNum = 0;
    int currFreq = 0;
    int maxFreq = 0;
    vector<int> result;
    void dfs(TreeNode* root) {
        if (!root) {
            return;
        }
        dfs(root->left);
        if (root->val == currNum) {
            currFreq++;
        } else {
            currNum = root->val;
            currFreq = 1;
        }
        if (currFreq > maxFreq) {
            result = {};
            maxFreq = currFreq;
        }
        if (currFreq == maxFreq) {
            result.push_back(root->val);
        }
        dfs(root->right);
    }
    vector<int> findMode(TreeNode* root) {
        dfs(root);
        return result;
    }
};

File: ADT/trees/501.py
class Solution:
    def __init__(self):
        self.mp = defaultdict(int)

    def dfs(self, root: Optional[TreeNode]):
        if not root:
            return
        self.dfs(root.left)
        self.mp[root.val] += 1
        self.dfs(root.right)

    def findMode(self, root: Optional[TreeNode]) -> List[int]:
        self.dfs(root)
        result = []
        maxFreq = 0
        for key, value in self.mp.items():
            if value > maxFreq:
                maxFreq = value
                result = [key]
            elif value == maxFreq:
                result.append(key)
        return result






class Solution:
    def __init__(self):
        self.currNum = 0
        self.currFreq = 0
        self.maxFreq = 0
        self.result = []

    def dfs(self, root: Optional[TreeNode]):
        if not root:
            return
        self.dfs(root.left)
        if root.val == self.currNum:
            self.currFreq += 1
        else:
            self.currNum = root.val
            self.currFreq = 1
        if self.currFreq > self.maxFreq:
            self.result = []
            self.maxFreq = self.currFreq
        if self.currFreq == self.maxFreq:
            self.result.append(root.val)
        self.dfs(root.right)

    def findMode(self, root: Optional[TreeNode]) -> List[int]:
        self.dfs(root)
        return self.result
